dimod/include/dimod/quadratic_model.h

```diff
@@ -837,118 +67,12 @@ class BinaryQuadraticModel : public QuadraticModelBase<Bias, Index> {
-            this->resize(bqm.num_variables());
-        }
-
-        // offset
-        this->offset() += bqm.offset();
-
-        // linear
-        for (size_type v = 0; v < bqm.num_variables(); ++v) {
-            base_type::linear(v) += bqm.linear(v);
-        }
-
-        // quadratic
-        for (size_type v = 0; v < bqm.num_variables(); ++v) {
-            if (bqm.adj_[v].size() == 0) continue;
-
-            this->adj_[v].reserve(this->adj_[v].size() + bqm.adj_[v].size());
-
-            auto span = bqm.neighborhood(v);
-            for (auto it = span.first; it != span.second; ++it) {
-                this->adj_[v].emplace_back(it->first, it->second);
-            }
-
-            this->adj_[v].sort_and_sum();
-        }
-    }
-
-    /**
-     * Add the variables, interactions and biases from another BQM.
-     *
-     * The size of the updated BQM will be adjusted appropriately.
-     *
-     * If the other BQM does not have the same vartype, the biases are adjusted
-     * accordingly.
-     *
-     * `mapping` must be a vector at least as long as the given BQM. It
-     * should map the variables of `bqm` to new labels.
-     */
-    template <class B, class I, class T>
-    void add_bqm(const BinaryQuadraticModel<B, I>& bqm, const std::vector<T>& mapping) {
-        if (bqm.vartype() != this->vartype()) {
-            // we could do this without the copy, but for now let's just do
-            // it simply
-            auto bqm_copy = BinaryQuadraticModel<B, I>(bqm);
-            bqm_copy.change_vartype(vartype());
-            this->add_bqm(bqm_copy, mapping);
-            return;
-        }
-
-        // offset
-        this->offset() += bqm.offset();
-
-        if (bqm.num_variables() == 0)
-            // nothing else to do, other BQM is empty
-            return;
-
-        // make sure we're big enough
-        T max_v = *std::max_element(mapping.begin(), mapping.begin() + bqm.num_variables());
-        if (max_v < 0) {
-            throw std::out_of_range("contents of mapping must be non-negative");
-        } else if ((size_type)max_v >= this->num_variables()) {
-            this->resize(max_v + 1);
-        }
-
-        // linear
-        for (size_type v = 0; v < bqm.num_variables(); ++v) {
-            this->linear(mapping[v]) += bqm.linear(v);
-        }
-
-        // quadratic
-        for (size_type v = 0; v < bqm.num_variables(); ++v) {
-            if (bqm.adj_[v].size() == 0) continue;
-
-            index_type this_v = mapping[v];
-
-            this->adj_[this_v].reserve(this->adj_[this_v].size() + bqm.adj_[v].size());
-
-            auto span = bqm.neighborhood(v);
-            for (auto it = span.first; it != span.second; ++it) {
-                this->adj_[this_v].emplace_back(mapping[it->first], it->second);
-            }
-
-            this->adj_[this_v].sort_and_sum();
-        }
+        this->add_quadratic_from_dense(dense, num_variables);
     }
 
     using base_type::add_quadratic;
```

Comment:
What does this mean ?

Reply:
Brings some of the overloads into this namespace.